src/token/wrap.rs

```diff
@@ -0,0 +1,290 @@
+            HostFunction::CreateTokenContractWithAsset,
+            parameters.into(),
+        )?;
+        let res_str = utils::vec_to_hash(&res)?;
+
+        let (storage, _, _) = h.try_finish().map_err(|_h| {
+            HostError::from(ScStatus::HostStorageError(
+                ScHostStorageErrorCode::UnknownError,
+            ))
+        })?;
+
+        snapshot::commit(state.1, ledger_info, &storage.map, &self.ledger_file).map_err(|e| {
+            Error::CannotCommitLedgerFile {
+                filepath: self.ledger_file.clone(),
+                error: e,
+            }
+        })?;
+        Ok(res_str)
+    }
+
+    async fn run_against_rpc_server(&self, asset: Asset) -> Result<String, Error> {
+        let client = Client::new(self.rpc_server_url.as_ref().unwrap());
+        let key = utils::parse_private_key(self.private_strkey.as_ref().unwrap())
+            .map_err(|_| Error::CannotParsePrivateKey)?;
+
+        // Get the account sequence number
+        let public_strkey =
+            stellar_strkey::StrkeyPublicKeyEd25519(key.public.to_bytes()).to_string();
+        // TODO: use symbols for the method names (both here and in serve)
+        let account_details = client.get_account(&public_strkey).await?;
+        // TODO: create a cmdline parameter for the fee instead of simply using the minimum fee
+        let fee: u32 = 100;
+        let sequence = account_details.sequence.parse::<i64>()?;
+        let contract_id = get_contract_id(&asset)?;
+        let tx = build_wrap_token_tx(
+            &asset,
+            &contract_id,
+            sequence,
+            fee,
+            self.network_passphrase.as_ref().unwrap(),
+            &key,
+        )?;
+
+        client.send_transaction(&tx).await?;
+
+        Ok(hex::encode(&contract_id))
+    }
+}
+
+fn get_contract_id(asset: &Asset) -> Result<Hash, Error> {
+    let preimage = HashIdPreimage::ContractIdFromAsset(asset.clone());
+    let preimage_xdr = preimage.to_xdr()?;
+    Ok(Hash(Sha256::digest(preimage_xdr).into()))
+}
+
+fn build_wrap_token_tx(
+    asset: &Asset,
+    contract_id: &Hash,
+    sequence: i64,
+    fee: u32,
+    network_passphrase: &str,
+    key: &ed25519_dalek::Keypair,
+) -> Result<TransactionEnvelope, Error> {
+    let lk = ContractData(LedgerKeyContractData {
+        contract_id: contract_id.clone(),
+        key: ScVal::Static(LedgerKeyContractCode),
+    });
+
+    let mut buf: Vec<u8> = vec![];
+    asset.write_xdr(&mut buf)?;
+    let parameters: VecM<ScVal, 256_000> =
+        vec![ScVal::Object(Some(ScObject::Bytes(buf.try_into()?)))].try_into()?;
+
+    let op = Operation {
+        source_account: None,
+        body: OperationBody::InvokeHostFunction(InvokeHostFunctionOp {
+            function: HostFunction::CreateTokenContractWithAsset,
+            parameters: parameters.into(),
+            footprint: LedgerFootprint {
+                read_only: VecM::default(),
+                read_write: vec![lk].try_into()?,
```

Comment:
`init_wrap` will also write `metadata`. It will also write an `admin` if the token is not native.